File: src/util/animation/Loop.java
package util.animation;

import java.util.function.LongConsumer;
import javafx.animation.AnimationTimer;

/**
 * Timer, that executes behavior in each frame while it is running.
 * <p/>
 * The methods {@link #start()} and {@link #stop()} allow to start and stop the timer.
 */
public class Loop {
	private boolean active;
	private final LongConsumer action;
	private final AnimationTimer timer = new AnimationTimer() {
		@Override
		public void handle(long l) {
			doLoop(l);
		}
	};

	/**
	 * Creates a new loop.
	 *
	 * @param action behavior to execute. Takes 1 parameter - The timestamp of the current frame given in nanoseconds.
	 * This value will be the same for all {@code AnimationTimers} called during one frame.
	 */
	public Loop(LongConsumer action) {
		this.action = action;
	}

	/** Creates a new loop. */
	public Loop(Runnable action) {
		this.action = now -> action.run();
	}

	/** Starts this loop. Once started, the behavior will be called in every frame. */
	public void start() {
		if (!active) {
			active = true;
			timer.start();
		}
	}

	/** Stops this loop. It can be activated again by calling {@link #start()}. */
	public void stop() {
		if (active) {
			timer.stop();
			active = false;
		}
	}

	/**
	 * This method needs to be overridden by extending classes. It is going to be called in every frame while the
	 * AnimationTimer is active.
	 *
	 * @param now The timestamp of the current frame given in nanoseconds. This value will be the same for all
	 * AnimationTimers called during one frame.
	 */
	protected void doLoop(long now) {
		action.accept(now);
	}
}

File: src/util/main/sp/it/util/animation/Loop.kt
package sp.it.util.animation

import javafx.animation.AnimationTimer
import sp.it.util.functional.invoke
import java.util.function.LongConsumer

/** Timer, that executes behavior in each frame while it is running. */
open class Loop {
    private var active: Boolean = false
    private val action: LongConsumer
    private val timer = object: AnimationTimer() {
        override fun handle(l: Long) {
            doLoop(l)
        }
    }

    /**
     * Creates a new loop with the specified action.
     *
     * @param action behavior to execute, that takes parameter - the timestamp of the current frame given in nanoseconds
     */
    constructor(action: LongConsumer) {
        this.action = action
    }

    /** Creates a new loop with the specified action. */
    constructor(action: Runnable) {
        this.action = LongConsumer { action() }
    }

    /** Starts this loop if it is not running. Once started, the action will be called in every frame. */
    fun start() {
        if (!active) {
            active = true
            timer.start()
        }
    }

    /** Stops this loop if it is running. */
    fun stop() {
        if (active) {
            timer.stop()
            active = false
        }
    }

    /**
     * @param now The timestamp of the current frame given in nanoseconds
     */
    protected open fun doLoop(now: Long) = action(now)

}